File: Day44/Part3/twin_dungeon.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <tuple>

using namespace std;

struct State {
    int rA, cA, rB, cB;
    int steps;
};

int main() {
    // Hardcoded Input
    vector<string> grid = {
        "#####",
        "#...#",
        "#.S.#",
        "#...#",
        "#####"
    };
    int N = grid.size();
    int M = grid[0].size();

    pair<int, int> startA = {1, 1};
    pair<int, int> goalA = {1, 3};
    pair<int, int> startB = {3, 3};
    pair<int, int> goalB = {3, 1};

    // BFS
    // dist[rA][cA][rB][cB]
    // Using flattened vector or just a visited array
    // Since N,M are small (5), we can use fixed array or vector
    // Let's use a 4D vector
    vector<vector<vector<vector<int>>>> dist(N, vector<vector<vector<int>>>(M, vector<vector<int>>(N, vector<int>(M, -1))));

    queue<State> q;
    q.push({startA.first, startA.second, startB.first, startB.second, 0});
    dist[startA.first][startA.second][startB.first][startB.second] = 0;

    int dr[] = {-1, 1, 0, 0};
    int dc[] = {0, 0, -1, 1};

    while (!q.empty()) {
        State cur = q.front();
        q.pop();

        if (cur.rA == goalA.first && cur.cA == goalA.second &&
            cur.rB == goalB.first && cur.cB == goalB.second) {
            cout << "Minimum steps: " << cur.steps << endl;
            return 0;
        }

        for (int i = 0; i < 4; ++i) {
            // Move A
            int nrA = cur.rA + dr[i];
            int ncA = cur.cA + dc[i];
            if (nrA < 0 || nrA >= N || ncA < 0 || ncA >= M || grid[nrA][ncA] == '#') {
                nrA = cur.rA;
                ncA = cur.cA;
            }

            // Move B
            int nrB = cur.rB + dr[i];
            int ncB = cur.cB + dc[i];
            if (nrB < 0 || nrB >= N || ncB < 0 || ncB >= M || grid[nrB][ncB] == '#') {
                nrB = cur.rB;
                ncB = cur.cB;
            }

            if (dist[nrA][ncA][nrB][ncB] == -1) {
                dist[nrA][ncA][nrB][ncB] = cur.steps + 1;
                q.push({nrA, ncA, nrB, ncB, cur.steps + 1});
            }
        }
    }

    cout << "-1" << endl;

    return 0;
}

File: learnLeetCode/Day44/Part3/twin_dungeon.py
import collections

def solve():
    # Hardcoded Input
    # 5x5 Grid
    # Walls: Border and center (2,2)
    grid = [
        "#####",
        "#...#",
        "#.S.#", # S is just a placeholder here, we use explicit coords
        "#...#",
        "#####"
    ]
    N = len(grid)
    M = len(grid[0])
    
    # Coordinates
    start_A = (1, 1)
    goal_A = (1, 3)
    
    start_B = (3, 3)
    goal_B = (3, 1)
    
    # BFS
    # State: (rA, cA, rB, cB)
    queue = collections.deque()
    queue.append((start_A[0], start_A[1], start_B[0], start_B[1], 0))
    
    visited = set()
    visited.add((start_A[0], start_A[1], start_B[0], start_B[1]))
    
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)] # U, D, L, R
    
    while queue:
        rA, cA, rB, cB, steps = queue.popleft()
        
        if (rA, cA) == goal_A and (rB, cB) == goal_B:
            print(f"Minimum steps: {steps}")
            return
        
        for dr, dc in directions:
            # Move A
            nrA, ncA = rA + dr, cA + dc
            if not (0 <= nrA < N and 0 <= ncA < M and grid[nrA][ncA] != '#'):
                nrA, ncA = rA, cA # Hit wall/bound, stay
            
            # Move B
            nrB, ncB = rB + dr, cB + dc
            if not (0 <= nrB < N and 0 <= ncB < M and grid[nrB][ncB] != '#'):
                nrB, ncB = rB, cB # Hit wall/bound, stay
                
            state = (nrA, ncA, nrB, ncB)
            if state not in visited:
                visited.add(state)
                queue.append((nrA, ncA, nrB, ncB, steps + 1))
                
    print("-1")

if __name__ == "__main__":
    solve()
